### PR TITLE
tests: branches coverage in `test_authentication`, `test_background`, `test_responses`, `test_routing`

### PR DESCRIPTION
**Pull request by lealre**
# Summary

Related to issue #2452

This PR addresses the cases in the test files below and essentially adds `#pragma: no branch` to all the cases, except for `test_authentication`, where I changed the `#pragma: no cover` line to cover the function that is not used.

From what I can see, there are no branch options to be covered, as the code relies on `elif` conditions or depends on a condition to pass the test.

| **File**                                                                                                       | **Branches**         |
| -------------------------------------------------------------------------------------------------------------- | -------------------- |
| [`tests/test_authentication.py`](https://github.com/encode/starlette/blob/master/tests/test_authentication.py) | 213->210             |
| [`tests/test_background.py`](https://github.com/encode/starlette/blob/master/tests/test_background.py)         | 76->exit             |
| [`tests/test_responses.py`](https://github.com/encode/starlette/blob/master/tests/test_responses.py)           | 268->exit, 752->exit |
| [`tests/test_routing.py`](https://github.com/encode/starlette/blob/master/tests/test_routing.py)               | 800->802, 874->exit  |

The only one I don't fully understand is `test_routing` ([line 874](https://github.com/encode/starlette/blob/master/tests/test_routing.py#L874)), where it's marking the @pytest.parametrized test as having uncovered branches.

# Checklist

- [x] I understand that this PR may be closed in case there was no previous discussion. (This doesn't apply to typos!)
- [ ] I've added a test for each change that was introduced, and I tried as much as possible to make a single atomic change.
- [ ] I've updated the documentation accordingly.

### CODE REVIEW

**Review by Kludex — APPROVED** [review]

File: tests/test_routing.py [hunk 1]
```diff
@@ -871,7 +871,7 @@ async def my_staticmethod(request: Request) -> None: ...  # pragma: no cover
     def __call__(self, request: Request) -> None: ...  # pragma: no cover
 
 
-@pytest.mark.parametrize(
+@pytest.mark.parametrize(  # pragma: no branch
```
Comment (Kludex) [inline tests/test_routing.py]: I don't know this one as well... 👀

Comment (Kludex) [reply tests/test_routing.py]: Can you try to find out why this works?

Comment (Kludex) [reply tests/test_routing.py]: We can merge the others here first, if you want.

Comment (lealre) [reply tests/test_routing.py]: Sure, I can try to find out as I will be working on the other cases.

I think we can merge the other ones if it's not a problem.

**Review by Kludex — APPROVED** [review]

File: tests/test_routing.py (hunk 1 above)
Comment (Kludex) [reply tests/test_routing.py]: Thanks.

Comment (Kludex) [reply tests/test_routing.py]:
```suggestion
@pytest.mark.parametrize(
```

**Pull request merged**